src/extras/reload.js

```diff
@@ -0,0 +1,69 @@
+(function (global) {
+  global.TRACING = true;
+
+  const systemJSPrototype = System.constructor.prototype;
+  const dependentsForModule = {};
+
+  systemJSPrototype.reload = function (id, parentUrl) {
+    const loader = this;
+
+    // An import first sets up a promise and resolves the module id, recreating
+    // the same behavior here avoids a race condition if a reload is called within
+    // that period
+    return Promise.resolve()
+      .then(function () {
+        return loader.resolve(id, parentUrl);
+      })
+      .then(function (id) {
+        const oldLoad = loader.getLoad(id);
+        if (!oldLoad) {
+          // module was not loaded before
+          return loader.import(id);
+        }
+
+        // module was imported before, and it might still be in a loading state
+        // first do a regular import for this module to ensure a current executing load
+        // is finished. except on error, as that will throw the same error again
+        const init = !oldLoad.er ? loader.import(id) : Promise.resolve();
+        return init
+          .catch(function () {
+            // ignore errors so that the reload can recover an error state
+          })
+          .then(function () {
+            // delete module from registry, and re-fetch it while maintaining importerSetters
+            loader.delete(id);
+
+            return loader.import(id, undefined, oldLoad.i).then(function (resolvedModule) {
+              const newLoad = loader.getLoad(id);
+
+              // replace references in dependants to the new module
+              if (id in dependentsForModule) {
+                dependentsForModule[id].forEach(function (dependent) {
+                  const dependentLoad = loader.getLoad(dependent);
+                  const indexOf = dependentLoad.d.indexOf(oldLoad);
+                  dependentLoad.d.splice(indexOf, 1, newLoad);
+                });
+              }
+
+              return resolvedModule;
+            });
+          });
+      });
+  }
+
+  const onload = systemJSPrototype.onload;
+  systemJSPrototype.onload = function onLoad(err, module, dependencies) {
+    if (!err && dependencies) {
```

Comment:
Note that errored modules may report dependencies if the error was an execution error and not a fetch error.

Reply:
I didn't consider execution errors at all, do these count as loading errors in native es modules too?

Reply:
Yes both are errors in this callback and on the load record, but the dynamic import behaviour in the browser is currently investigating retrying fetch errors.